src/plugins/intel_gna/src/gna_plugin.cpp

```diff
@@ -898,6 +899,8 @@ void GNAPlugin::LoadNetwork(const CNNNetwork& _network) {
     if (sortedNet.empty()) {
         THROW_GNA_EXCEPTION << "Sorted network is empty";
     }
+    // put Assign (sortedNet) layers at the end to properly update states variables (Read)
+    std::stable_partition(sortedNet.begin(), sortedNet.end(), [&](CNNLayerPtr layer){return !LayerInfo(layer).isAssign();});
```

Comment:
This is not right, they executes after matching state variable without fix.
How about that:
// Copy operations connected to memory layer (Assign to state variable) should be executed when all functional layers are calculated.

Reply:
I'am fine with this.
Yor description is also correct and better than the previous one.
I would only change "when" to "after", ie:
// Copy operations connected to memory layer (Assign to state variable) should be executed **after** all functional layers are calculated.

However, IMO the main reason to execute Copy->Memory after functional is that "something" is reading this Copy destination buffer after Copy executes (without the fix), and AFAIK "something" must be Memory counterpart which is ReadValue.

Without the fix:

1. F1 part executes
2. Copy->Memory executes, and the data is copied to State buffer S1
3. F2 part executes
4. S1 is read (due to ReadValue connected to some input of F3) and some related F3 part executes but this in not correct
5. F4 part executes (this is independent of S1)

So we need to move 2 after 4 (not necessarily after 5)
I am fine to move 2 after 5 since it is also correct.
so maybe:

// Copy operations connected to memory layer (Assign to state variable X) must be executed **after** functional operations dependent on X are executed. To simplify, just moving these Copy operations at the end of the execution list.

Reply:
We have no steps 2 and 4 at all, that is the reason why we faced with the issue. F1 and F4 connected with the same "Read memory" buffer. So this is no a simplifying, this is how it should work. All states buffers should be assigned when inference is finished but to minimize copy operation we put this logic to gna model (using the same memory pointer for read and assign buffers)

Reply:
Updated comment

Reply:
OK, resolving